singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,374 @@
+//------------------------------------------------------------------------------
+// © 2021-2022. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_VINET_HPP_
+#define _SINGULARITY_EOS_EOS_VINET_HPP_
+#define SINGULARITY_ENABLE_EXCEPTIONS
+
+#include <cmath>
+#include <cstdio>
+
+#include <limits>
+
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/base/math_utils.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+
+// COMMENT: This is a complete, thermodynamically consistent Vinet
+// EOS created at Sandia National Laboratories by AEM.
+// J. Appl. Phys. 119, 015904 (2016)
+// It is particularly suited for high-pressure applications.
+class Vinet : public EosBase<Vinet> {
+ public:
+  Vinet() = default;
+  // Constructor
+  PORTABLE_INLINE_FUNCTION
+  Vinet(const Real rho0, const Real T0, const Real B0, const Real BP0, const Real A0,
+        const Real Cv0, const Real E0, const Real S0, const Real *expconsts)
+      : _rho0(rho0), _T0(T0), _B0(B0), _BP0(BP0), _A0(A0), _Cv0(Cv0), _E0(E0), _S0(S0),
```

Comment:
`std::vector` is also automatically cleaned up, i.e., no need to call `free`. The only issue is it doesn't work on device. But if the constructor is only called on host, that's fine. We can pass in `std::vector`, and then convert into portable memory later when you define `_d2to40` and `_VIP`.

Reply:
> I second @jhp-lanl . I _really_ don't like passing a raw pointer here. I see that you're converting this to static memory later, but this is super unsafe and absolutely asking for segfaults and memory errors.

If it's just a `view` pointer (no memory manipulation) there isn't much to be fret about. From what I recall there isn't any actual difference between using `type *` and `type []`, and in fact it's a strict syntax fault to declare an array parameter as explicitly sized. If the concern is about conforming to style, then I would argue we need to redo `Real* lambda`.

`_expconsts` looks to is just be used to initialize `_d2to40`, if so, it doesn't need to be carried around.

Reply:
`Real *lambda` is deliberately flexible and also works on GPU. I find that distasteful as well, but there's a reason for it. I would rather not do that here unless there's a good reason, especially since this is a CPU-side constructor. I don't see any reason not to pass in something like `std::vector` here.

> _expconsts looks to is just be used to initialize _d2to40, if so, it doesn't need to be carried around.

Yes agreed.